docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,958 @@
+> Additionally:
+> - The MachineHealthCheck controller adds the HealthCheckSucceeded and the OwnerRemediated conditions.
+> - The KubeadmControlPlane adds the APIServerPodHealthy, ControllerManagerPodHealthy, SchedulerPodHealthy, EtcdPodHealthy, EtcdMemberHealthy conditions.
+
+Notes:
+- This proposal introduces a mechanism for extending the meaning of Machine readiness, `ReadinessGates` (see [changes to Machine.Spec](#machine-spec)).
+- While `Ready` is the main signal for machines operational state, higher level abstractions in Cluster API like e.g. 
+  MachineDeployment are relying on the concept of Machine's `Availability`, which can be seen as readiness + stability.
+  In order to standardize this concept across different higher level abstractions, this proposal is surfacing `Availability`
+  condition at Machine level as well as adding a new `MinReadySeconds` field (see [changes to Machine.Spec](#machine-spec))
+  that will be used to compute this condition.
+- Similarly, this proposal is standardizing the concept of Machine's `UpToDate` condition, however in this case it will be up to
+  the Machine's owner controllers to set this condition.
+- Conditions like `NodeReady` and `NodeHealthy` which depend on the connection to the remote cluster will benefit
+  from the new `RemoteConnectionProbe` condition at cluster level (see [Cluster (New)Conditions](#cluster-newconditions));
+  more specifically those condition should be set to `Unknown` after the cluster probe fails
+  (or after whatever period is defined in the `--remote-conditions-grace-period` flag)
+- `HealthCheckSucceeded` and `OwnerRemediated` (or `ExternalRemediationRequestAvailable`) conditions are set by the 
+  MachineHealthCheck controller in case a MachineHealthCheck targets the machine.
+- KubeadmControlPlane also adds additional conditions to Machines, but those conditions are not included in the table above
+  for sake of simplicity (however they are documented in the KubeadmControlPlane paragraph).
+
+TODO: think carefully at remote conditions becoming unknown, this could block a few operations ... 
+
+#### Machine Spec
+
+Machine's spec is going to be improved to allow 3rd party components to extend the semantic of the new Machine's `Ready` condition
+as well as to standardize the concept of Machine's `Availability`.
+
+Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
+Below the Go types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.
+
+```go
+type MachineSpec struct {
+
+    // MinReadySeconds is the minimum number of seconds for which a Machine should be ready before considering the replica available.
+    // Defaults to 0 (machine will be considered available as soon as the Node is ready)
+    // +optional
+    MinReadySeconds int32 `json:"minReadySeconds,omitempty"`
```

Comment:
A user wouldn't create the Machine, but the MD. But I guess the question remains the same (especially as it defaults to 0)

A good thing, it can be configured in the ClusterClass

Reply:
Might be what is not clearly evident above, is the fact that `MinReadySeconds` field already exists in Cluster API MachineDeployments and MachinePools, this proposal is just surfacing this field at Machine level with the goal to allow users looking at the output of `kubectl get machines` to have have the full set of info behind the availability condition. 

WRT using `readinessGates` to express "wait for something specific" instead of waiting arbitrarily for a few seconds is one of the possible use cases for `readinessGates`, but this requires users to implement a custom controller that adds a custom condition to Machine's conditions.

Having `MinReadySeconds` can help users not willing to implement a custom controller, even if I agree that just adding some delay might be racy in most cases.

Another thing to consider, and most probably also the reason why `MinReadySeconds` field exists in CAPI today, is the fact that the same field exists in Kubernetes DaemonSets, Deployments, StatefulSet. 
Having `MinReadySeconds` also in Cluster API resources might help users approaching the project

TL;DR this proposal is not adding this feature, just making its usage more explicit.
`readinessGates` provides an alternative to handle more sophistical wait strategies.

Reply:
> Having MinReadySeconds can help users not willing to implement a custom controller, even if I agree that just adding some delay might be racy in most cases.

Not sure I understand how MinReadySeconds could be racy

Otherwise agree.

I think MinReadySeconds is just an easy way to ensure Available doesn't flicker if Ready is flickering (which can easily happen because of e.g. MemoryPressure condition that is included

Reply:
Resolving, hopefully it is clear now